commons-stack/praise
# Feat/user event log

Resolves #309 
Resolves #375 

**Overview**
- new api endpoint `/api/eventlogs/all`
- helper utility `logEvent()`
- tests on api endpoint & helper utility
- Frontend for viewing list of events, paginated
- Add `logEvent()` calls everywhere that a user event occurs (see list above)
- Refactor logic to determine a user's display name: move from frontend to backend, attach to UserDto as `nameRealized`

To limit the scope for this PR, I did not include custom filtering / searching / sorting of event logs. (I created another ticket #376). I also did not include infinite-scroll. **Let me know if you think its preferable to just build out that functionality as part of this PR.**

--

Our outline of this feature in #309 was kinda fuzzy, so I tried to think it through and decide on what to include in this UserEventLog data. This is all open to feedback and revision! 

Here's what I came up with:
- The purpose of this log is *user-facing community transparency*, not systems administration nor debugging.
- For that purpose, the only important data is "changes that someone made to the database state"
- Thus, user-interaction commands which do not change the database state should **not** be logged.
- Thus, user actions which "fail", and do not change the database state, should **not** be logged

So the only data included in this log are the following **user actions which change the database state**:
- Log in
- Create Period
- Update Period
- Close Period
- Assign random quantifiers to all praise in period
- Update Period Setting
- Update Global Setting
- Add role to user
- Remove role from user
- Quantify praise
- Run /activate command on discord (initiate UserAccount activation)
- Create Forwarded Praise
- Create Praise
- Replace quantifiers (cli command)
- Import praise (cli command)

This does not account for changes made by a sysadmin directly to the database itself (i.e. bypassing the application interfaces).

**Also, just a note that currently makes individual quantification scores public, before the quantification period is closed.** Perhaps it should be modified to only include quantification user events after the related period is closed.
